### PR TITLE
New: Allow All Directives in Line Comments

### PR DESCRIPTION
**Pull request by captbaritone**
_Note: This PR was previously titled "Warn When Directives Use the Wrong Type of Comment" but has evolved into a different proposal_

## Summary

ESLint currently only allows the following types in directives to be used inside of block comments (`/* Block Comment */`):

* `exported`
* `global(s)`
* `eslint-(en|dis)able`
* `eslint`

Currently if a user adds one of these directives in a line comment (`// Line Comment`) the directive will be silently ignored. Which can be very confusing.

I propose that we allow all directives to work in either type of comment.

## Related Issues

* [Warn when ESLint directives which expect to be used in block level comments are used in inline comments #12014](https://github.com/eslint/eslint/issues/12014)

### CODE REVIEW

**Comment by ljharb** [comment]
Could perhaps, instead of a warning (which is better than silent failure), it Just Work? Why does the comment style matter?

**Comment by captbaritone** [comment]
Great question, I’m afraid I don’t fully know. I assumed it was to ensure those more serious directives were given a weightier visual appearance, but @kaicataldo mentioned in https://github.com/eslint/eslint/issues/12014 that it might be about enforcing some kind of granularity. Any idea who would know what the intention behind this limitation was?

**Comment by captbaritone** [comment]
I went searching through the Git and GitHub histories and tried to see if I could find any clues as to why we make a distinction between inline comments and block comments. I didn't find anything too concrete, but my instinct tells me that defining globals and enableing/disabling rules per-file were copied from JSHint which used block comments for this purpose.

I didn't see any good reason to leave this convention in place, and I saw a number of good reasons to get rid of it.

I would say @platinumazure and @ilyavolodin seem likely to have context on this.

Here are some interesting things I found, in chronological order:

## Add no-undef (fixes #6) [#164](https://github.com/eslint/eslint/pull/164)

_Jul 28, 2013_

This PR adds the `/* global */` syntax. The requirement that comments be block level is not questioned in the PR, but I get the impression that the choice was borrowed from JSHint/JSLint.

## Disabling/enabling rules through comments [#320](https://github.com/eslint/eslint/pull/320)

_Oct 3, 2013_

In a [comment](https://github.com/eslint/eslint/pull/320#discussion_r7234133) nzakas questions why only block comments are supported.

@ilyavolodin cites a convention used in JSHint that block comments are for file-level configuration and inline are used for per-line config.

nzakas replies (emphasis mine):

> Hmm okay. I'm not sure I want to emulate the JSHint behavior if this is true, **it seems like block for global and non-block for local is too implicit to have meaning.**

But agrees to punt.

## New: Adding Support for "// eslint-disable-line rule" style comments [#1840](https://github.com/eslint/eslint/pull/1840)

_Feb 16, 2015_

In a [comment](https://github.com/eslint/eslint/pull/1840#issuecomment-74612637) nzakas suggests these rules be required to be inline. The PR author pushes back citing consistency and stating that all the other comments use block style. They eventually opt for requiring that `eslint-disable-line` use an inline comment.

This is where we start to have two code paths, one for inline and one for block level.

## Docs: clarify usage of inline disable comments (fixes #6335) [#6347](https://github.com/eslint/eslint/pull/6347)

_Jun 8, 2016_

@kaicataldo updates the docs to clarify which comment directives require block level comments. In the related issue @platinumazure comments:

> Yes, we very intentionally allow certain inline config comments to be line-only, and others are block-only. We could probably improve the documentation on that front.

But does not provide any context as to why.

**Comment by kaicataldo** [comment]
I'm interested to hear from some other team members who might have more context as to why we limit certain inline config comments to block comments, but I'm also unsure why we can't support both. Seems like that would be the best developer experience.

**Comment by ilyavolodin** [comment]
The guess that globals were added as block level comments to comply with JSHint is correct. The idea back then was that JSHint is a lot more popular, and a lot of people probably already have comments in their files declaring globals. If they wanted to switch to ESLint they would have to create a new configuration file, but at least, they wouldn't have to go file by file to change global variable declarations for linter.
I'm not opposed to supporting both styles of comments for all directives, but I do think that directives that deal with enabling/disabling things probably make more sense as inline comments vs block (disable next line, for example).

**Comment by kaicataldo** [comment]
That's right, I'm remembering the discussion around allowing block comments for `eslint-disable-line` and `eslint-disable-next-line` comments now. The issue we ran into was with how multiline block comments would be treated - i.e. how do you know which line should be disabled? As far as I can tell, this isn't an issue with other comments that aren't based on their line location.

**Comment by ljharb** [comment]
I'd assume the next line after the comment block ends or the line on which the comment block ends, and I'm not clear on what other possibilities there'd be?

**Comment by not-an-aardvark** [comment]
We already support multiline `eslint-disable-[next-]line` comments and report an error if the comment spans multiple lines. Changing all directives to support either line or block comments seems fine to me.

**Comment by captbaritone** [comment]
Okay. It seems like the idea of allowing any directive to exist in any type of comment has no immediate objections. Should I edit this RFC to reflect this new plan of attack, or open a new one?

**Comment by platinumazure** [comment]
I think editing this RFC should be fine.

**Comment by captbaritone** [comment]
I've updated the RFC to reflect the new plan of attack.

Note that I've switched to using "line comments" to refer to `// comment`, since this matches the AST syntax and I have no idea if "inline comment" is a clear name.

**Comment by mysticatea** [comment]
How about `/* eslint-env */` comments? It's on [lib/linter/linter.js#L413](https://github.com/eslint/eslint/blob/fb08b7c9d28bc68864eb940e26df274059228b6a/lib/linter/linter.js#L413) because ESLint must find it before parsing.

**Comment by captbaritone** [comment]
@mysticatea I don't see any reason we couldn't extend that regex to support Line comments. If nobody objects, I'll add `eslint-env` to the RFC as well.

File: designs/2019-allow-all-directives-in-line-comments/README.md
Comment (platinumazure) [inline designs/2019-allow-all-directives-in-line-comments/README.md]:
```suggestion
An earlier revision of this RFC suggested reporting an error/warning when a directive that is only supported in a block comment was found in a line comment.
```

File: designs/2019-allow-all-directives-in-line-comments/README.md
Comment (platinumazure) [inline designs/2019-allow-all-directives-in-line-comments/README.md]: My spellcheck was flagging this one but it also flagged the way I would normally type it, so I came up with a longer way-- disregard if you like.

```suggestion
As a user of ESLint I occasionally wish to disable a rule, declare a valid global or add some other file level config. In doing so I often forget that these directives are only valid inside block level comments and add them as inline comments by mistake. When the comment fails to have its desired effect I consult the documentation to see what I did wrong. Did I remember the directive syntax or name incorrectly? On more than one occasion the problem has been that I used the wrong kind of comment.
```

**Comment by ljharb** [comment]
I would be shocked to discover a single instance in any code anywhere of a directive-like comment that was not intended to be a directive.

**Comment by platinumazure** [comment]
@ljharb An example in ESLint's own codebase was identified in the RFC. The comment was talking about the `globalReturn` parser option but the comment started with `// global return`.

**Comment by ljharb** [comment]
@platinumazure i guess for globals comments, since `global` is allowed (even though i only ever use `globals`), but what about every other kind of directive?

**Comment by platinumazure** [comment]
> but what about every other kind of directive?

I can't think of any examples of those, outside of maybe "eslint" (e.g., a comment `// eslint is being silly here` and we interpret that as trying to set an `is` rule).

**Comment by captbaritone** [comment]
@platinumazure I like the idea of trying to limit what we interpret as directives in order to reduce the likelihood of accidentally over-matching. However, I don't think forcing the directive to be alone on a line will work.

I'm guessing your hesitancy was due to these same concerns, but I'll spell them out just in case:

* It's very common (and useful) to use the rest of the line of the directive to explain the "why" behind the directive.
* Ignoring malformed directives, as opposed to matching them and triggering an error, can lead to the same type of bad developer experience that this RFC is designed to mitigate. If a user misunderstands the subtle details of how to construct a directive, it simply doesn't work rather than resulting in an error message which can help them discover their error.

Another option would be to write an ESlint rule which could detect these latent/accidental directives so that users could explicitly handle them before upgrading. While a bit cute, this solution is probably not great because it asks quite a lot of the person performing the upgrade.

1. Install this random ESLint rule
2. Resolve/address all issues
3. Remove the rule
4. Upgrade

I guess another option would be to go back to the original version of this RFC: Warn when we detect inline directives. And get that merged into a minor version and then wait some time before merging the change to make them allowed.

Honestly, that seems like a bit much given the low likelihood of these types of comments.

**Comment by platinumazure** [comment]
@captbaritone I think I understand your point of view. I'm still no closer to having a good idea how to resolve the concerns. Hoping other team members will be able to weigh in.

**Comment by captbaritone** [comment]
I'm not sure how to drive this forward. Who would be able to decide if we are willing to turn existing line comments that are accidental directives into actual directives?

**Comment by platinumazure** [comment]
@captbaritone Thinking on it again, I think I'm okay with this as long as we release this with a major version and include it in the migration guide. Is that an acceptable compromise?

Also, thanks for following up!

**Comment by captbaritone** [comment]
That sounds fair. How frequent are major releases? If the next one won't be for a while (many months) it might be worth considering:

* In a minor version release make line comment directives a warning so that devs know why they aren't working and can fix any accidental directives before the major version release.
* In the next major version start allowing line comment directives and remove the previous warnings.

However, if the next major version will be relatively soon, the value of that extra work becomes pretty small.

**Comment by kaicataldo** [comment]
@captbaritone We generally do one major release a year sometime in the first half of the year, so our next major release is a little ways off at this point.

**Comment by platinumazure** [comment]
@captbaritone Could you update the RFC with the note that the full change is a breaking change, along with a detailed proposal about the warnings (along with a note that they may only be needed if the breaking change is far off)? Thanks!

**Comment by ilyavolodin** [comment]
@platinumazure Can you elaborate as to why you think this is a breaking change? I understand that if somebody used inline comments that were not supported before, they would now start seeing errors. However, I would guess that if they used inline comments that was a mistake and unintentional. So this might be considered a bug fix that increases number of errors which would require a minor version bump, but I'm not sure it rises to the level or a breaking change and requires major version.

**Comment by captbaritone** [comment]
@ilyavolodin As called out in the RFC, under "Accidental Directives" there was at least one example in the ESLint code base (https://github.com/eslint/eslint/blob/02d7542cfd0c2e95c2222b1e9e38228f4c19df19/lib/rules/newline-before-return.js#L137) where there was a line comment

> `// global return at beginning of script`

Which after this change would be newly interpreted as a directive. It's conceivable that other code baseses would have similar comments which ESLint would treat as directives but were not intended to be interpreted as such.

**Comment by ilyavolodin** [comment]
Ah, yes. Thanks for pointing that out. That would be a breaking change.

**Comment by captbaritone** [comment]
Cool. I'll proceed to update the RFC as suggested by @platinumazure.

**Comment by captbaritone** [comment]
Updated.

File: designs/2019-allow-all-directives-in-line-comments/README.md
Comment (ilyavolodin) [inline designs/2019-allow-all-directives-in-line-comments/README.md]: Can you update this section with the explanation that those changes will be backwards-compatible but would be considered breaking due to the example with `// global` comment.

File: designs/2019-allow-all-directives-in-line-comments/README.md
Comment (ilyavolodin) [inline designs/2019-allow-all-directives-in-line-comments/README.md]: Another possibility would be to make the changes to all comments but `// global` for now, and leave global to the next major release.

Comment (platinumazure) [reply designs/2019-allow-all-directives-in-line-comments/README.md]: I don't think `// global` is the only example. If someone writes `// eslint is stupid`, then that will attempt to be parsed as a configuration of the rule `is` (and will fail due to lack of colon).

Comment (ilyavolodin) [reply designs/2019-allow-all-directives-in-line-comments/README.md]: Fair enough. I'll remove this comment.

File: designs/2019-allow-all-directives-in-line-comments/README.md
Comment (kaicataldo) [inline designs/2019-allow-all-directives-in-line-comments/README.md]: Small typo: `types in directives` -> `types of directives`

**Comment by captbaritone** [comment]
In https://github.com/eslint/eslint/pull/12397 @mysticatea brought up the fact that maybe even adding warnings when we detect directives in line comments should be considered a breaking change since will potentially add new errors when people upgrade.

If we think that's true, that leaves two options:

1. Go back to the original version of this RFC and just warn on directives in line comments. Ship that in a major release.
2. Continue with the current approach outlined in this RFC except spread it over two major releases instead of a minor release and a major release.

If we _don't_ think adding these errors is a breaking change, then we can continue with this RFC as currently proposed.

**Comment by ljharb** [comment]
Warnings aren’t a breaking change unless they change the exit code.

**Comment by platinumazure** [comment]
Another possibility might be to skip the warning stage and just introduce the breaking change in the next major release.

Or, another possibility might be to create a new CLI option or supplemental tool to detect these, allow users to use that option starting from a minor release, and then do the breaking change in the major release.

The issue is that we can't increase warnings by default (i.e., not behind some type of backward compatible option) in a minor release.

**Comment by ljharb** [comment]
If it’s not reported as an error, why not?

**Comment by platinumazure** [comment]
> If it’s not reported as an error, why not?

Because of `eslint --max-warnings=0`, which would result in an error/nonzero exit code if we add a warning by default.

I would love to see our semver policy changed to carve that out as an exception (similar to how using `eslint:all` naturally requires a little more care when upgrading and we have an exception to the semver policy in that case). But right now, I don't think we can do this in a semver-minor change.

That said, the TSC is able to change the semver policy (and in fact, we met today and will be announcing a change). So that's another option: change the policy, then allow more warnings (but not errors) on semver-minor releases.

**Comment by ljharb** [comment]
It makes sense to me to allow that as an exception, since the user using the flag has opted into it.

**Comment by captbaritone** [comment]
Maybe I should just publish a separate package which contains a lint rule which warns/errors on block-only directive appearing in line comments. Then we can ship enabling line comment directives in the next major version and just link to the separate package/rule in the changelog.

This allows us to get there in one major version and also allows for a smooth upgrade path. The downside is that the smooth upgrade path will be a bit annoying to take (have to read the docs to find out about installing a new package and enabling a new rule just for the transition period, and then delete/disable it).

The other downside is that if you _don't_ find out about the external package which provides the smooth the upgrade path, the failures you experience will be very confusing since they will manifest as malformed directives: adding a bunch of random words to your global scope, or setting a bogus environment.

**Comment by mysticatea** [comment]
I should mention our Node.js API such as `CLIEngine` or `Linter`. Warnings don't trigger non-zero exit code in our CLI. But I'm not sure for Node.js API users.

**Comment by platinumazure** [comment]
> Warnings don't trigger non-zero exit code in our CLI. But I'm not sure for Node.js API users.

API users are using those classes within a Node process, so it's up to them how they want to exit. I think from an ESLint perspective, adding more warnings wouldn't be a breaking change to them. It's only a breaking change in CLI because of `--max-warnings`.

**Comment by mysticatea** [comment]
> API users are using those classes within a Node process, so it's up to them how they want to exit.

Exactly. That's the reason that I think increasing warnings for enhancement is a breaking change.

**Comment by platinumazure** [comment]
> That's the reason that I think increasing warnings for enhancement is a breaking change.

I don't understand your reasoning. API users can do whatever they want with more or fewer warnings, including ignore them entirely. Why would adding warnings be a breaking change for API users?

**Comment by mysticatea** [comment]
Your explanation is the reason. It's regular use cases of APIs. Enhancement should not break regular use cases.

**Comment by platinumazure** [comment]
In that case, should we treat warnings as exactly the same as errors from an API perspective? Should we convert all warnings into errors for API users? What is the point of having the theoretically lower severity at all, if we treat them the same?

I think of it the other way around: We have warnings to allow users to have lint messages without breaking their lint build (unless they specifically change that behavior with `--max-warnings`). We should assume API users have the same understanding and not worry about extra warnings (as long as they are always only warnings and never errors).

Happy to put this in a TSC meeting issue if needed. We're getting off track here.

@captbaritone Any thoughts on where you would like to go with this? I think current semver policy is that any increase in warnings requires a major release.

Do you want to wait for TSC to discuss potentially changing the semver policy around warning-only messages, or do you think it makes more sense to just wait for this change to happen in our next major release?

Currently, we're tentatively aiming to start major release planning in November with a tentative goal of being completely through betas and RCs and releasing the actual major release by the end of January at the earliest.

**Comment by mysticatea** [comment]
> In that case, should we treat warnings as exactly the same as errors from an API perspective?

When we consider if an enhancement needs semver-major or not, yes.

> Should we convert all warnings into errors for API users?

No. The warning severity has another role rather than that we avoid semver-mejor in enhancements.

> What is the point of having the theoretically lower severity at all, if we treat them the same?

API users can distinguish warnings (including the rules which are configured `severity=1`) and errors if they want.

> I think of it the other way around: We have warnings to allow users to have lint messages without breaking their lint build (unless they specifically change that behavior with --max-warnings). We should assume API users have the same understanding and not worry about extra warnings (as long as they are always only warnings and never errors).

I don't object that we do so, but I think it's a breaking change and we should update our documentation.

**Comment by captbaritone** [comment]
Assuming adding the warning _and_ enabling line level directives are each breaking changes, my temptation would be to do each one in a major version.

Honestly my original proposal was just to add the warnings since the real problem I set out to solve was the improve the developer experience when a user mistakenly uses a line comment where only a block comment is permitted.

Getting a warning instead of silently ignoring the directive at least allows the user to quickly fix their code.

**Comment by platinumazure** [comment]
I've created this [TSC meeting agenda item](https://github.com/eslint/tsc-meetings/issues/153#issuecomment-546118064) to try to resolve the semver question around warnings.

**Comment by platinumazure** [comment]
The TSC [met yesterday and discussed this](https://github.com/eslint/tsc-meetings/blob/master/notes/2019/2019-12-05.md#can-adding-warnings-without-adding-errors-ever-be-semver-minor), and we agree, in principle, that adding warnings to improve UX for at least this use case could be allowable as a semver-minor change. Apologies for the delay in discussing this.

I've been tasked with creating an issue in eslint/eslint to discuss exactly how we want to do it, so we still have a few things to figure out. But in principle, we should be able to convert this RFC back to warning on line comments (if desired) and getting it out in semver-minor change.

My proposal is roughly going to be this:

1. Short-term, no technical changes, just process changes. On a case-by-case basis, TSC can allow specific exceptional cases where new warnings (and no errors) are generated and we will release semver-minor. These exceptional cases will *never* involve rules giving more warnings-- that will always be semver-minor if a bug fix only, or semver-major if not a bug fix. (And we would then exercise this process to allow this RFC to move forward, hopefully.)
1. Long-term, we will create a new property on the lint result object to contain configuration warnings or errors (or, more generally, warnings/errors that apply to the whole lint run rather than individual files/rules), and configuration warnings will never count against the `--max-warnings` CLI flag. During major releases, we would migrate appropriate file-level warnings into this property as necessary.

I understand this has been a slow and frustrating process and I greatly appreciate your patience with us as we work through this. Thank you @captbaritone for your continued willingness to work with us on this RFC!

**Comment by platinumazure** [comment]
Earlier, I had proposed that another option might be to create a new CLI option or supplemental tool which lets users *opt in* to checking for these. This might be the best solution if we want to allow users to continue to write real comments (e.g., `// eslint is buggy here` immediately before an `eslint-disable-next-line` comment), but still give users a way to potentially find unexpectedly not-working ESLint comments.

To me, this completely bypasses the discussion about adding warnings and semver-major/-minor.

Thoughts?

**Comment by btmills** [comment]
We discussed this in today's TSC meeting as a potential breaking change for the v7.0.0 release. Since we have the option of a process change that would allow us to include this in a semver-minor release, we'd prefer to do that, and we don't have to hurry to get this in as part of the v7.0.0 release. More details on the process change can be found in https://github.com/eslint/eslint/issues/12650.

**Comment by glen-84** [comment]
Recent comments seem to imply that single-line comment directives will not be supported (they will only result in a warning)?

Single-line comment directives are supported in stylelint and PHP_CodeSniffer, so please consider supporting them in ESLint.

I don't see why this would be a breaking change.

**Comment by nzakas** [comment]
@eslint/eslint-tsc it looks like we lost track of this RFC. How do we want to proceed?

**Comment by mdjermanovic** [comment]
After reading the RFC document and all discussions, I'm not sure that I understand what's the final proposal.

* Warn on directive-like line comments in a minor version, and treat them as real directives in a major version?
* Warn on directive-like line comments in a major version, and treat them as real directives in a subsequent major version?
* Just warn on directive-like line comments in a minor version?
* Just warn on directive-like line comments in a major version?
* Just treat them as real directives in a major version?

**Comment by glen-84** [comment]
Could `exported` be changed to `eslint-exported`, and `global` to `eslint-global`?

They don't seem very unique currently.

**Comment by ljharb** [comment]
I believe `global` (and `globals`) both long predate eslint; it seems like it would cause a lot of disruption to change them.

**Comment by glen-84** [comment]
Are they used by other tools? The `eslint-` prefixed versions could be added as alternatives initially.

Anyway, just a thought.

**Comment by ljharb** [comment]
By other linters, absolutely - not sure about other tools beyond that.

**Comment by glen-84** [comment]
I don't think that it matters if other linters (f.e. jshint) use the same directive, as a project would only use one linter.

If you switched linters, you'd have to update other directives like `*-disable` etc. anyway.

**Comment by ljharb** [comment]
Nowadays, I agree, but I used to run all of jshint, jscs, and eslint on hundreds of projects all at once.

**Comment by nzakas** [comment]
@glen-84 this is off topic. If you’d like to discuss further, please create a new discussion. Let’s keep these comments targeted at the RFC.

**Comment by nzakas** [comment]
@mdjermanovic the RFC is proposing to introduce warnings in a minor version and then allow all directives in both block and line comments in a major version. The intent is to warn people about comments that look like directives but do nothing before we actually make them do something.

The warnings part is considered optional, though, so we could just skip directly to making directives work in all comments in a major version.

**Review by mdjermanovic — APPROVED** [review]

**Comment by glen-84** [comment]
> 
> 
> @glen-84 this is off topic. If you’d like to discuss further, please create a new discussion. Let’s keep these comments targeted at the RFC.

Is it off topic? It's listed in the RFC under `Accidental Directives`.

By allowing `global` (for example) in line comments, it may cause issues with existing or even new code. That is why I was suggesting moving to a prefixed version of these directives.

Anyway, I apologize for the noise. I'll hide my previous messages.

**Comment by nzakas** [comment]
I understand your point, it’s just moving to prefixed directives is beyond the scope of this RFC at this point. It’s already in the final commenting period, during which we are mostly making a go/no-go decision.

**Comment by nzakas** [comment]
Okay, we like the direction, so merging. @captbaritone our apologies for losing track of this. We don’t have any excuse for letting this slip beneath our radar for so long.

**Pull request merged**
